Update for bare metal area

iriark01 opened the pull request:
There are questions commented out in the sources. There are some things I didn't know.

iriark01 committed:
Update for bare metal area

There are questions commented out in the sources. There are some things I didn't know.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,19 @@
-<span class="images">![Mbed OS bare metal profile block diagram](../../images/bare_metal_block_diagram.png)<span>Mbed OS bare metal profile block digram</span></span>
+Bare metal doesn't use the RTOS APIs. Instead, it relies on timers to control the workflow.
+<!--not sure I get it - I looked at both Blinky examples and they both use thread_sleep_for, even though the bare metal one doesn't include mbed_thread.h-->
+<!--what other APIs does it have or not have by default?-->
```
Bare metal uses a subset of the RTOS APIs. These APIs have been ported to bare metal i.e. they do not make calls to RTX. By default, bare metal includes: drivers, platform and rtos-api https://github.com/ARMmbed/mbed-os/blob/master/platform/bare_metal/mbed_lib.json/#L3

evedon commented on docs/program-setup/bare_metal/bare_metal_api.md:
```diff
@@ -1,47 +1,26 @@
+By default, bare metal includes the following APIs:
 
-<h2 id="digital-i-o">Digital I/O</h2>
+## Adding APIs
```
This will be covered in the tutorial. This page shoud only list what is supported and the default list of APIs.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,19 @@
+The Mbed OS tools - Mbed CLI, Mbed Online Compiler and Mbed Studio all support working with the bare metal profile.
 
-The Mbed OS bare metal profile is a compact profile of Mbed OS without an RTOS. The Mbed OS bare metal profile is API compatible with Mbed OS and supports a subset of Mbed OS features, such as analog I/O, digital I/O, timers and digital interfaces, such as SPI, I2C, Serial and CAN. Mbed Studio, Mbed CLI and the Mbed Online Compiler all support the Mbed OS bare metal profile.
+Note that Mbed TLS and Mbed Crypto are not supported for bare metal.
```
They are both supported. Rajkumar is in the process of updating this table.

evedon commented on docs/program-setup/bare_metal/c_small_libs.md:
```diff
@@ -1,5 +1,7 @@
 <h1 id="using-small-c-libraries">Using small C libraries in Mbed OS bare metal</h1>
-
+<!--is this page for app developers? do we expect them to always manually specify for their app that they want to use bare metal?-->
+<!--I'm also intriggued by the phrasing "if your application doesn't use an RTOS". Of course it doesn't; we're talking about bare metal-->
```
OK- sentence needs rephrasing. The point was to say "1- use bare metal profile if you don't need a RTOS 2- use small library with bare metal profile"

iriark01 replied:
So if I have to use this in bare metal apps, should it be part of - or at least mentioned - in the new tutorial in PR https://github.com/ARMmbed/mbed-os-5-docs/pull/1305?

evedon commented on docs/program-setup/bare_metal/c_small_libs.md:
```diff
@@ -1,5 +1,7 @@
 <h1 id="using-small-c-libraries">Using small C libraries in Mbed OS bare metal</h1>
-
+<!--is this page for app developers? do we expect them to always manually specify for their app that they want to use bare metal?-->
```
At the moment the default for all targets is full Mbed OS. So users/applications need to explicitly enable the bare metal profile. This should change post Mbed 6.

evedon committed:
Added Mbed 2 to Mbed OS bare metal porting guide (#1293)

* Added Mbed 2 to Mbed OS bare metal porting guide

* Apply suggestions from code review

* Tidying up

I will take another look at it on the preview branch to make sure all formatting works

* Wording fix

Co-authored-by: Irit Arkin <irit.arkin@arm.com>

iriark01 committed:
Adding bare metal porting to JSON

iriark01 committed:
First table - just to see what it looks like

iriark01 committed:
Replacing Gap class link

iriark01 committed:
Giving UX some symbols to choose from

iriark01 committed:
Better preview for UX

iriark01 committed:
A bit more for testing

iriark01 committed:
Index is ready

iriark01 committed:
More content from PR 1304

iriark01 committed:
Some tidying up from comments

iriark01 committed:
Quick edit

iriark01 committed:
Comments in intro

iriark01 committed:
Invalid JSON

iriark01 committed:
Spans broke the table

iriark01 committed:
Missed some spans

iriark01 committed:
Draft for review

This includes all I have at the moment

iriark01 committed:
Tidying up

iriark01 committed:
Trying another place for the pic

iriark01 committed:
Typo fix

iriark01 committed:
Will this work?

iriark01 committed:
Removing diagram

evedon commented on docs/api/api.md:
```diff
@@ -1,346 +1,259 @@
+| - | - | - |
+| [Thread](../apis/rtos-apis.html) | &#10004; | |
+| [ThisThread](../apis/thisthread.html) | &#10004; | &#10004; (enabled by default) |
+| [Mutex](../apis/mutex.html) | &#10004; | Partial |
```
I actually like the two-column approach. 

There is no need for `Partial` any longer:
- In the case of Mutex we should say that it is supported (the fact that it is a dummy implementation does not matter for the purpose of this page).
- In the case of Kernel interface functions, let's just say that only `get_ms_count` is supported (the other functions don't make sense in a bare metal application)

evedon commented on docs/api/api.md:
```diff
@@ -1,346 +1,259 @@
+| [EventFlags](../apis/eventflags.html) | &#10004; | &#10004; |
+| [Event](../apis/event.html) | &#10004; | &#10004; |
+| [Conditionvariable](../apis/conditionvariable.html) | &#10004; | |
+| [Kernel interface functions](../apis/kernel-interface-functions.html) | &#10004; | Partial |
```
@rajkan01 Please review this page and correct what is enabled by default.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,231 @@
-| Mbed PSA | Not available | Available |
-
-To begin using the Mbed OS bare metal profile from Mbed OS 2, please follow our [instructions](../tutorials/migrating-to-mbed-os-5.html).
+Bare metal is a profile of Mbed OS for ultraconstrained hardware. It represents a different way of working with Mbed OS: Instead of enabling all features by default, as the full profile does, bare metal mandates only the smallest set of APIs that development boards require - driver APIs, platform APIs and a subset of the RTOS APIs. All other supported APIs are optional - you can pull them in if y ...
```
Before we say that it is a different way of working with Mbed OS, we should say that it is a compact profile of Mbed OS without an RTOS. 

I would say : "bare metal mandates only the smallest set of APIs that applications require ..."

I would leave the final sentence "but you'll probably leave most of them out - reducing your application's final size.". Instead, perhaps say that you can better control your application's final size.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,231 @@
-To begin using the Mbed OS bare metal profile from Mbed OS 2, please follow our [instructions](../tutorials/migrating-to-mbed-os-5.html).
+Bare metal is a profile of Mbed OS for ultraconstrained hardware. It represents a different way of working with Mbed OS: Instead of enabling all features by default, as the full profile does, bare metal mandates only the smallest set of APIs that development boards require - driver APIs, platform APIs and a subset of the RTOS APIs. All other supported APIs are optional - you can pull them in if y ...
+
+Bare metal uses a subset of Mbed OS's RTOS APIs, without becoming a full RTOS. It's therefore suitable for applications that do not require complex thread management. This not only simplifies your application code, it also means you can use APIs that are not thread safe. Just as important, you can use the code-optimized versions of the C standard libraries, `microlib` and `newlib-nano`, which are ...
```
Proposition for first sentence : "The bare metal profile implements a subset of Mbed OS's RTOS APIs that are useful in a non threaded application, for example semaphores (with release from interrupts) or tickers (to set up a recurring interrupt).

iriark01 replied:
What does "with release from interrupts" mean? Specifically the "with".

evedon replied:
You would not want to acquire a semaphore in interrupt context but releasing a semaphore is perfectly fine. So you could say `(calling the release API from interrupts)`.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,231 @@
+
+Bare metal uses a subset of Mbed OS's RTOS APIs, without becoming a full RTOS. It's therefore suitable for applications that do not require complex thread management. This not only simplifies your application code, it also means you can use APIs that are not thread safe. Just as important, you can use the code-optimized versions of the C standard libraries, `microlib` and `newlib-nano`, which are ...
+
+The bare metal profile is determined at build time - your working environment includes the full Mbed OS library, but only the bare metal APIs are compiled.
```
This sentence is not clear. We show in the tutorial how to use the bare metal profile.

iriark01 replied:
I was trying to say that people shouldn't go looking for a second download of Mbed OS - they'll always get the same download, but the application will be built differently based on whether or not they enable bare metal.

evedon replied:
I think it is implied when we say that bare metal is a profile that `uses a subset of Mbed OS's RTOS APIs`.

iriark01 replied:
I'm not sure. In the same way that DM Client and DM Client Lite are two different downloads, not two different builds of the same download.

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,231 @@
+        <tr>
+            <td rowspan="6">Core</td>
+            <td>Drivers</td>
+            <td>Available; enabled by default <br>(Except the class `USBCDC_ECM`)</td>
```
The formatting is not clear. 
Also, I think that we give too much detail here by listing the one exception (this will be ported to bare metal soon).
(and we should use "API" instead of "class")

evedon commented on docs/program-setup/bare_metal/bare_metal.md:
```diff
@@ -1,22 +1,231 @@
+        </tr>
+            <tr>        
+            <td>RTOS APIs</td>
+            <td>Semaphore, Mutex, EventFlags, ThisThread: Available; enabled by default.<br><br> All other classes not enabled.</td>
```
Same comment about formatting. Since you show on the second page what is enabled by default I would remove that information here.
```suggestion
            <td>Semaphore, Mutex, EventFlags, ThisThread: Available; enabled by default.<br><br> All other APIs are not supported.</td>
```

evedon commented on docs/program-setup/bare_metal/bare_metal_api.md:
```diff
@@ -1,5 +1,33 @@
 # Bare metal APIs
```
This page will be deleted, won't it?

iriark01 replied:
Yes, but I didn't want to remove it from the branch until we knew we had all the info we needed from it. It's not shown in the docs, only on GitHub.

evedon commented on docs/program-setup/bare_metal/c_small_libs.md:
```diff
@@ -58,5 +63,5 @@ After you have completed the steps above, add `small` to the `supported_c_libs`
 
 ### Note on uARM toolchain
-
+<!--I think all refrences to uArm should be removed at this point - including the scatter file bits above-->
```
We need to keep the scatter file information somewhere as it is very useful.

iriark01 replied:
I'm just worried about the uArm bit, since we deprecated it.

evedon replied:
But we have not removed all the TOOLCHAIN_ARM_MICRO folders in the code base, specifically for this reason.

iriark01 replied:
Oh, I didn't know that. So this page is still 100% accurate?

iriark01 committed:
Updated from Evelyne's comments

iriark01 committed:
Merge remote-tracking branch 'origin/bare-metal-work' into bare-metal-work

iriark01 committed:
Merge branch 'development' into bare-metal-work

iriark01 committed:
Quick starts to offer both Blinkys

iriark01 commented:
I think I need a clarification for usage: bare metal "isn't an RTOS" or "doesn't have an RTOS"? Or "isn't an RTOS and doesn't have RTOS APIs"?

iriark01 committed:
Not showing LoRAWAN as supported

iriark01 committed:
Trying a new version of the API index

iriark01 committed:
Changing tick for manually enabled

iriark01 committed:
Demonstrating page order

iriark01 committed:
Util > Platform

iriark01 committed:
Is this a better title?

iriark01 committed:
New attempt at APIs

This is now putting three tasks together - bare metal API indications, new API structure, and deprecated API list.

iriark01 committed:
Driver tidying

iriark01 committed:
Function link as tip

iriark01 committed:
Updating intro

iriark01 committed:
Intro text update

iriark01 committed:
Adding Wait

iriark01 committed:
Adding Wait.md so the branch can build

iriark01 committed:
Fixing deprecated APIs 

Thanks, Evelyne!

iriark01 committed:
Format fixes

iriark01 commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+
+## Deprecated APIs: moving from Mbed OS 5 to 6
+
+If you're moving your program from Mbed OS 5 to 6, you will need to replace deprecated APIs:
```
```suggestion
If you're moving your program from Mbed OS 5 to 6, you will need to replace deprecated APIs. The table lists classes that have been completely removed. Functions and methods that have been removed from other classes are listed in each class's Doxgen, and [summarised here](../mbed-os-api-doxy/deprecated.html).
```

iriark01 commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+| `InterruptManager` | No replacement. |
+| `CallChain` | No replacement. |
+
+<span class="tips">The full Doxygen offers [a list of deprecated functions](../mbed-os-api-doxy/deprecated.html).</span>
```
```suggestion
```

evedon commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+| [USBAudio](../apis/usbaudio.html) | &#10004; | &#10004;  |
+| [USBHID](../apis/ubshid.html) | &#10004; | &#10004;  |
+| [USBCDC](../apis/usbcdc.html) | &#10004; | &#10004;  |
+| [USBCDC_ECM](../apis/usbcdc-ecm.html) | &#10004; | &#10004;  |
```
This is currently not supported in bare metal. We have a task to port it. It should be done by next week.

evedon commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+| - | - | - |
+| [Power management (sleep)](../apis/power-apis.html) | &#10004; | &#10004; |
+| [DeepSleepLock](../apis/deepsleeplock.html) | &#10004; | &#10004;  |
+| [Idle loop](../apis/idle-loop.html) | &#10004; | &#10004;  |
```
This belongs to RTOS section (although it is related to power maangement since by default the idle loop invokes sleep manager to enter a sleep mode). Header and source files are under rtos/ so let's keep the documentation at the same place. Also this is NOT supported in bare metal (again confusion because the API was moved from rtos to platform table and so does not match the code)

evedon commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+
+### Time
+
+| API | Full profile | Bare metal profile |
```
Could you give the API column a minimum width?

iriark01 replied:
I'm afraid not - we don't control the way the tables are rendered

evedon commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+| [Error handling](../apis/error-handling.html) | &#10004; | &#10004;  |
+| [Assert](../apis/assert.html) | &#10004; | &#10004;  |
+| [NonCopyable](../apis/noncopyable.html) | &#10004; | &#10004;  |
+| [SharedPtr (Shared pointer](../apis/shared-pointer.html) | &#10004; | &#10004;  |
```
A parenthesis is missing

evedon commented on docs/api/api_alternative.md:
```diff
@@ -0,0 +1,289 @@
+| BLE services: `iBeacon`, `UARTService`, `URIBeaconConfigService` | No replacement available |
+| `TCPServer` | `TCPSocket` |
+| `NVStore` | `KVStore` |
+| `FunctionPointerArg1<R, A>` & `FunctionPointerArg1<R, void>` | `Callback<R(A)>` & `Callback<void()>` |
```
I provided the list for completeness but I would list only is there was a .md file for the API. Can you check if this one and those below were mentioned in 5.15?

iriark01 replied:
"FunctionPointerArg1" not in the docs
"RtosTimer" mentioned in Tickless and Execution
"InterruptManager" not in the docs
"CallChain" not in the docs

I'm more concerned with "will they be able to understand why their code doesn't compile" than "explicitly had a page"

evedon replied:
I propose to remove "FunctionPointerArg1" and keep the others then.

evedon commented on docs/api/platform/Wait.md:
```diff
@@ -0,0 +1,18 @@
+# Wait
+
+<span class="tips">**Tip:** `wait` is deprecated in favor of explicit sleep functions. To sleep, replace `wait` with `ThisThread::sleep_for` (C++) or `thread_sleep_for` (C). To wait (without sleeping), call `wait_us`. `wait_us` is safe to call from ISR context.</span>
```
It is more than a tip?
`wait` is not only deprecated but also removed in Mbed 6

iriark01 replied:
Didn't we just put it back? Did I misunderstand PR https://github.com/ARMmbed/mbed-os-5-docs/pull/1306?

iriark01 replied:
Or am I confusing a class with a function/method?

evedon replied:
We have removed the `wait` function but we have kept the `wait_us` and `wait_us` functions.

evedon commented on docs/bare_metal/bare_metal_api.md:
```diff
@@ -1,5 +1,33 @@
 # Bare metal APIs
 
+The bare metal profile starts with a restricted list of API, to which you can add what you need. However, bare metal doesn't support all Mbed OS APIs.
```
I am assuming that this page will be deleted.

iriark01 replied:
Yes.

evedon commented on docs/api/api.md:
```diff
@@ -1,343 +1,253 @@
 # Full API list
-<!--these are now just the APIs - do I want to add the tutorials?-->
+
```
I am assuming that this page is replaced by the alternative.

iriark01 replied:
Yes

evedon commented on docs/bare_metal/bare_metal.md:
```diff
@@ -0,0 +1,108 @@
+
+Bare metal is a profile of Mbed OS for ultraconstrained hardware: compact and without an RTOS. It represents a different way of working with Mbed OS: the bare metal profile builds only the smallest set of APIs that applications require - driver APIs, platform APIs and a subset of the RTOS APIs. This gives you better control of the application's final size than the full profile, which relies on th ...
+
+The bare metal profile implements a subset of Mbed OS's RTOS APIs that are useful in non-threaded applications, such as semaphores (calling the release API from interrupts) and tickers (to set up a recurring interrupt). It does not include an RTX and is not a full RTOS, and is therefore suitable for applications that do not require complex thread management. Instead of the RTOS's scheduler, all a ...
```
"and is not a full RTOS" : I would remove. We've already said in the first sentence that it is "without an RTOS"

evedon commented on docs/bare_metal/bare_metal.md:
```diff
@@ -0,0 +1,108 @@
+        </tr>
+        <tr>
+            <td>BLE</td>
+            <td>Available<br>(Except on `TARGET_NORDIC_CORDIO`)</td>
```
For completeness, currently two drivers are not supported: COMPONENT_BlueNRG_MS and TARGET_NORDIC_CORDIO. We should either list which ones are supported and which ones aren't or say that not all drivers are supported? FI, the cordio driver could be ported to bare metal next week.

iriark01 replied:
If there are only two we don't support, I think listing both is nice, as it removes the question "well if you say 'not all drivers' how do I know which ones".

iriark01 replied:
```suggestion
            <td>Available<br>Except `COMPONENT_BlueNRG_MS` and `TARGET_NORDIC_CORDIO`</td>
```

evedon replied:
Good (but the highlighter does not work)

iriark01 replied:
Okay, I'll figure the formatting out

iriark01 committed:
Some comments from Evelyne + moving files

iriark01 committed:
Organising Driver APIs

iriark01 committed:
Platform APIs

iriark01 committed:
JSON fix + trying new format for table

iriark01 committed:
I think this tidies up all APIs

iriark01 committed:
Differentiated marks (UX request)

iriark01 committed:
Tidying up the lists

This includes alphabetising them. It splits apart some related APIs - like BufferedSerial and UnbufferedSerial - but it makes it easier to scan the list.

iriark01 committed:
Regrouping APIs

Will alphabetise the ToC when I'm sure of the groupings

iriark01 committed:
Everything alphabetised

iriark01 committed:
Final comments
